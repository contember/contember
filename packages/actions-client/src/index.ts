export * from './runtimes'
export * from './payloadHandler'
export * from './requestHandler'
export * from './runtimeHandler'
export * from './types'
