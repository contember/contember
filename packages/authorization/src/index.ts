export * from './Authorizator'
export * from './AccessEvaluator'
export * from './AccessNode'
export * from './AuthorizationScope'
export * from './Permissions'
