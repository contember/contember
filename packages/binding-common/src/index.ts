export * from './markers'
export * from './treeParameters'
export * from './environment'
export * from './queryLanguage'
export * from './schema'
export * from './types'

export * from './consts'
export * from './BindingError'
export * from './LocalizedBindingError'
export * from './RuntimeId'
export * from './utils/TreeNodeUtils'

export { v4 as generateUuid } from 'uuid'
