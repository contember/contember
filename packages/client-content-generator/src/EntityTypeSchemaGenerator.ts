import { Model } from '@contember/schema'
import { acceptEveryFieldVisitor, acceptFieldVisitor } from '@contember/schema-utils'
import { getEnumTypeName } from './utils'

export type GenerateOptions = {
	includeDeprecated?: boolean
}

export class EntityTypeSchemaGenerator {
	generate(model: Model.Schema, options?: GenerateOptions): string {
		let code = ''
		for (const enumName of Object.keys(model.enums)) {
			code += `import type { ${getEnumTypeName(enumName)} } from './enums'\n`
		}

		code += `
export type JSONPrimitive = string | number | boolean | null
export type JSONValue = JSONPrimitive | JSONObject | JSONArray
export type JSONObject = { readonly [K in string]?: JSONValue }
export type JSONArray = readonly JSONValue[]

`

		const entities = options?.includeDeprecated ? Object.values(model.entities) : Object.values(model.entities).filter(entity => !entity.deprecationReason)
		for (const entity of entities) {
			code += this.generateTypeEntityCode(model, entity, options)
		}
		code += '\n'
		code += `export type ContemberClientEntities = {\n`
		for (const entity of entities) {
			code += `\t${entity.name}: ${entity.name}\n`
		}
		code += '}\n\n'
		code += `export type ContemberClientSchema = {\n`
		code += '\tentities: ContemberClientEntities\n'
		code += '}\n'
		return code
	}

	private generateTypeEntityCode(model: Model.Schema, entity: Model.Entity, options?: GenerateOptions): string {
		let code = `export type ${entity.name} <OverRelation extends string | never = never> = {\n`
		code += '\tname: \'' + entity.name + '\'\n'
		code += '\tunique:\n'
		code += this.formatUniqueFields(model, entity)
		let columnsCode = ''
		let hasOneCode = ''
		let hasManyCode = ''
		acceptEveryFieldVisitor(model, entity, {
			visitHasMany: ctx => {
				if (!options?.includeDeprecated && ctx.relation.deprecationReason) {
					return
				}
				hasManyCode += `\t\t${ctx.relation.name}: ${ctx.targetEntity.name}${ctx.targetRelation?.type === Model.RelationType.ManyHasOne ? `<'${ctx.targetRelation.name}'>` : ''}\n`
			},
			visitHasOne: ctx => {
				if (!options?.includeDeprecated && ctx.relation.deprecationReason) {
					return
				}
				hasOneCode += `\t\t${ctx.relation.name}: ${ctx.targetEntity.name}\n`
			},
			visitColumn: ctx => {
				if (!options?.includeDeprecated && ctx.column.deprecationReason) {
					return
				}
				columnsCode += `\t\t${ctx.column.name}: ${columnToTsType(ctx.column)}${ctx.column.nullable ? ` | null` : ''}\n`
			},
		})

		code += '\tcolumns: {\n'
		code += columnsCode
		code += '\t}\n'
		code += '\thasOne: {\n'
		code += hasOneCode
		code += '\t}\n'
		code += '\thasMany: {\n'
		code += hasManyCode
		code += '\t}\n'
		code += '\thasManyBy: {\n'
		code += this.formatReducedFields(model, entity)
		code += '\t}\n'
		code += '}\n'
		return code
	}

	private formatReducedFields(model: Model.Schema, entity: Model.Entity, options?: GenerateOptions): string {
		let code = ''
		acceptEveryFieldVisitor(model, entity, {
			visitOneHasMany: ({ entity, relation, targetEntity, targetRelation }) => {
				if (!options?.includeDeprecated && relation.deprecationReason) {
					return
				}
				if (!targetRelation) {
					return
				}
				if (!options?.includeDeprecated && targetRelation.deprecationReason) {
					return
				}
				const uniqueConstraints = getFieldsForUniqueWhere(model, targetEntity)
				const composedUnique = uniqueConstraints
					.filter(fields => fields.length === 2) //todo support all uniques
					.filter(fields => fields.includes(targetRelation.name))
					.map(fields => fields.filter(it => it !== targetRelation.name))
					.map(fields => fields[0])
				const singleUnique = uniqueConstraints
					.filter(fields => fields.length === 1 && fields[0] !== targetEntity.primary)
					.map(fields => fields[0])
					.filter(it => it !== targetRelation.name)

				;[...composedUnique, ...singleUnique].forEach(fieldName => {
					const capitalizeFirstLetter = (value: string) => {
						return value.charAt(0).toUpperCase() + value.slice(1)
					}
					const name = `${relation.name}By${capitalizeFirstLetter(fieldName)}`

					const targetUnique = targetEntity.fields[fieldName]

					code += `\t\t${name}: { entity: ${targetEntity.name}; by: {${fieldName}: ${uniqueType(model, targetEntity, targetUnique)}}  }\n`

				})
			},
			visitColumn: () => {
			},
			visitManyHasManyInverse: () => {
			},
			visitManyHasManyOwning: () => {
			},
			visitManyHasOne: () => {
			},
			visitOneHasOneInverse: () => {
			},
			visitOneHasOneOwning: () => {
			},
		})
		return code
	}

	private formatUniqueFields(model: Model.Schema, entity: Model.Entity): string {
		const fields = getFieldsForUniqueWhere(model, entity)
		let code = ''
		for (const field of fields) {
			code += '\t\t| Omit<{ '
			code += field.map(it => `${it}: ${uniqueType(model, entity, entity.fields[it])}`).join('; ')
			code += '}, OverRelation>\n'
		}
		return code
	}
}


const uniqueType = (model: Model.Schema, entity: Model.Entity, field: Model.AnyField): string => {
	return acceptFieldVisitor(model, entity, field, {
		visitColumn: ctx => {
			return columnToTsType(ctx.column)
		},
		visitRelation: ctx => {
			return ctx.targetEntity.name + `['unique']`
		},
	})
}


const columnToTsType = (column: Model.AnyColumn): string => {
	const baseType = (() => {
		switch (column.type) {
			case Model.ColumnType.Enum:
				return getEnumTypeName(column.columnType)
			case Model.ColumnType.String:
				return 'string'
			case Model.ColumnType.Int:
				return 'number'
			case Model.ColumnType.Double:
				return 'number'
			case Model.ColumnType.Bool:
				return 'boolean'
			case Model.ColumnType.DateTime:
				return 'string'
			case Model.ColumnType.Time:
				return 'string'
			case Model.ColumnType.Date:
				return 'string'
			case Model.ColumnType.Json:
				return 'JSONValue'
			case Model.ColumnType.Uuid:
				return 'string'
			default:
				((_: never) => {
					throw new Error(`Unknown type ${_}`)
				})(column.type)
		}
	})()
	return column.list ? `readonly ${baseType}[]` : baseType
}

const getFieldsForUniqueWhere = (schema: Model.Schema, entity: Model.Entity): readonly (readonly string[])[] => {
	const relations = Object.values(
		acceptEveryFieldVisitor<undefined | [string]>(schema, entity, {
			visitColumn: () => undefined,
			visitManyHasManyInverse: () => undefined,
			visitManyHasManyOwning: () => undefined,
			visitOneHasMany: ({ relation }) => [relation.name],
			visitManyHasOne: () => undefined,
			visitOneHasOneInverse: ({ relation }) => [relation.name],
			visitOneHasOneOwning: ({ relation }) => [relation.name],
		}),
	).filter((it): it is [string] => !!it)

	return [[entity.primary], ...Object.values(entity.unique).map(it => it.fields), ...relations]
}
