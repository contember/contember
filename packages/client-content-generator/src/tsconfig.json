{
	"extends": "../../../tsconfig.settings.json",
	"compilerOptions": {
		"outDir": "../dist/types",
		"types": ["node"]
	},
	"references": [
		{ "path": "../../cli-common/src" },
		{ "path": "../../client-content/src" },
		{ "path": "../../schema/src" },
		{ "path": "../../schema-utils/src" },
	]
}
