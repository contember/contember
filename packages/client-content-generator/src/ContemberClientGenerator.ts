import { Model } from '@contember/schema'
import { EnumTypeSchemaGenerator } from './EnumTypeSchemaGenerator'
import { EntityTypeSchemaGenerator, GenerateOptions } from './EntityTypeSchemaGenerator'
import { NameSchemaGenerator } from './NameSchemaGenerator'


export class ContemberClientGenerator {
	constructor(
		private readonly nameSchemaGenerator: NameSchemaGenerator = new NameSchemaGenerator(),
		private readonly enumTypeSchemaGenerator: EnumTypeSchemaGenerator = new EnumTypeSchemaGenerator(),
		private readonly entityTypeSchemaGenerator: EntityTypeSchemaGenerator = new EntityTypeSchemaGenerator(),
	) {
	}

	generate(model: Model.Schema, options?: GenerateOptions): Record<string, string> {
		const nameSchema = this.nameSchemaGenerator.generate(model)
		const enumTypeSchema = this.enumTypeSchemaGenerator.generate(model)
		const entityTypeSchema = this.entityTypeSchemaGenerator.generate(model, options)

		const namesCode = `import { SchemaNames } from '@contember/client-content'
export const ContemberClientNames: SchemaNames = ` + JSON.stringify(nameSchema, null, 2)

		const indexCode = `import { ContemberClientNames } from './names'
import type { ContemberClientSchema } from './entities'
import { ContentQueryBuilder, TypedContentQueryBuilder, TypedEntitySelection } from '@contember/client-content'
export * from './names'
export * from './enums'
export * from './entities'

export const queryBuilder = new ContentQueryBuilder(ContemberClientNames) as unknown as TypedContentQueryBuilder<ContemberClientSchema>

export type FragmentOf<EntityName extends keyof ContemberClientSchema['entities'] & string, Data = unknown> =
TypedEntitySelection<ContemberClientSchema, EntityName, ContemberClientSchema['entities'][EntityName], Data>

export type FragmentType<T extends TypedEntitySelection<any, any, any, any> = any> =
T extends TypedEntitySelection<any, any, any, infer TFields>
	? TFields
	: never
`
		return {
			'names.ts': namesCode,
			'enums.ts': enumTypeSchema,
			'entities.ts': entityTypeSchema,
			'index.ts': indexCode,
		}
	}
}
