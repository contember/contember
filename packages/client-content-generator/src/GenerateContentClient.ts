import * as fs from 'node:fs/promises'
import { resolve, join } from 'node:path'
import { Command, CommandConfiguration, Input } from '@contember/cli-common'
import { ContemberClientGenerator } from './ContemberClientGenerator'

type Args = {
	['schema-path']?: string
	['out-dir']?: string
}

type Options = {
	['include-deprecated']?: boolean
}

const usageInfo = `
Usage:

From file:
yarn contember-client-generator <schema.json> <out-dir> [--include-deprecated]

From stdin:
yarn run --silent contember project:print-schema --format=schema | yarn contember-client-generator - <out-dir> [--include-deprecated]
`

export class GenerateContentClient extends Command<Args, Options> {

	protected configure(configuration: CommandConfiguration<Args, Options>) {
		configuration.description('Generate content client.')

		configuration.argument('schema-path').optional().description('Path to the json schema file. Tip: Use "-" to read from stdin.')
		configuration.argument('out-dir').optional().description('Output directory for generated files.')

		configuration.option('include-deprecated').valueNone().description('Include deprecated entities and fields')
	}

	protected async execute(input: Input<Args, Options>) {
		const schemaPath = input.getArgument('schema-path')
		const outDir = input.getArgument('out-dir')
		const includeDeprecated = input.getOption('include-deprecated')

		if (!schemaPath || !outDir) {
			// eslint-disable-next-line no-console
			console.info(usageInfo)
			throw 'Missing required arguments <schema.json> and <out-dir>'
		}

		const sourceData = await (async () => {
			if (schemaPath === '-') {
				if (process.stdin.isTTY) {
					throw 'Cannot read from stdin in TTY mode'
				}
				const buffer = []
				for await (const chunk of process.stdin) {
					buffer.push(chunk)
				}
				return Buffer.concat(buffer).toString('utf8')
			}
			return await fs.readFile(resolve(process.cwd(), schemaPath), 'utf8')
		})()

		const source = JSON.parse(sourceData)

		const dir = resolve(process.cwd(), outDir)
		const generator = new ContemberClientGenerator()
		const result = generator.generate(source.model, { includeDeprecated })

		await fs.mkdir(dir, { recursive: true })

		for (const [name, content] of Object.entries(result)) {
			await fs.writeFile(join(dir, name), content, 'utf8')
		}
	}
}
