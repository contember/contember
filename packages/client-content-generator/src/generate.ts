#!/usr/bin/env node
import { GenerateContentClient } from './GenerateContentClient'

(async () => {
	await new GenerateContentClient().run(process.argv.slice(2))
})().catch(e => {
	console.error(e)
	process.exit(1)
})
