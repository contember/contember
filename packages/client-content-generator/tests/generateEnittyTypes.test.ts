import { describe, expect, test } from 'bun:test'
import { EntityTypeSchemaGenerator } from '../src'
import { schemas } from './schemas'


describe('generate entities', () => {

	const entityGenerator = new EntityTypeSchemaGenerator()
	test('generate for scalars', () => {

		expect(entityGenerator.generate(schemas.scalarsSchema.model)).toMatchSnapshot()
	})

	test('generate for scalars with deprecated', () => {

		expect(entityGenerator.generate(schemas.scalarsSchema.model, { includeDeprecated: true })).toMatchSnapshot()
	})

	test('generate for has enum', () => {

		expect(entityGenerator.generate(schemas.enumSchema.model)).toMatchSnapshot()
	})

	test('generate for has enum deprecated', () => {

		expect(entityGenerator.generate(schemas.enumSchema.model, { includeDeprecated: true })).toMatchSnapshot()
	})


	test('generate one has one', () => {

		expect(entityGenerator.generate(schemas.oneHasOneSchema.model)).toMatchSnapshot()
	})

	test('generate one has one deprecated', () => {

		expect(entityGenerator.generate(schemas.oneHasOneSchema.model, { includeDeprecated: true })).toMatchSnapshot()
	})

	test('generate one has many', () => {

		expect(entityGenerator.generate(schemas.oneHasManySchema.model)).toMatchSnapshot()
	})

	test('generate one has many deprecated', () => {

		expect(entityGenerator.generate(schemas.oneHasManySchema.model, { includeDeprecated: true })).toMatchSnapshot()
	})

	test('generate many has many', () => {

		expect(entityGenerator.generate(schemas.manyHasManySchema.model)).toMatchSnapshot()
	})

	test('generate many has many deprecated', () => {

		expect(entityGenerator.generate(schemas.manyHasManySchema.model, { includeDeprecated: true })).toMatchSnapshot()
	})

	test('generate reduced has by', () => {

		expect(entityGenerator.generate(schemas.reducedHasManySchema.model)).toMatchSnapshot()
	})

	test('generate reduced has by deprecated', () => {

		expect(entityGenerator.generate(schemas.reducedHasManySchema.model, { includeDeprecated: true })).toMatchSnapshot()
	})

})
