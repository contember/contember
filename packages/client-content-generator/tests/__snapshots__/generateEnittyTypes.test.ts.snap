// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`generate entities generate for scalars 1`] = `
"
export type JSONPrimitive = string | number | boolean | null
export type JSONValue = JSONPrimitive | JSONObject | JSONArray
export type JSONObject = { readonly [K in string]?: JSONValue }
export type JSONArray = readonly JSONValue[]

export type Foo <OverRelation extends string | never = never> = {
	name: 'Foo'
	unique:
		| Omit<{ id: string}, OverRelation>
	columns: {
		id: string
		stringCol: string | null
		intCol: number | null
		doubleCol: number | null
		dateCol: string | null
		datetimeCol: string | null
		booleanCol: boolean | null
		jsonCol: JSONValue | null
		uuidCol: string | null
	}
	hasOne: {
	}
	hasMany: {
	}
	hasManyBy: {
	}
}

export type ContemberClientEntities = {
	Foo: Foo
}

export type ContemberClientSchema = {
	entities: ContemberClientEntities
}
"
`;

exports[`generate entities generate for scalars with deprecated 1`] = `
"
export type JSONPrimitive = string | number | boolean | null
export type JSONValue = JSONPrimitive | JSONObject | JSONArray
export type JSONObject = { readonly [K in string]?: JSONValue }
export type JSONArray = readonly JSONValue[]

export type Foo <OverRelation extends string | never = never> = {
	name: 'Foo'
	unique:
		| Omit<{ id: string}, OverRelation>
	columns: {
		id: string
		stringCol: string | null
		intCol: number | null
		doubleCol: number | null
		dateCol: string | null
		datetimeCol: string | null
		booleanCol: boolean | null
		jsonCol: JSONValue | null
		uuidCol: string | null
		deprecatedStringCol: string | null
		deprecatedIntCol: number | null
		deprecatedDoubleCol: number | null
		deprecatedDateCol: string | null
		deprecatedDatetimeCol: string | null
		deprecatedBooleanCol: boolean | null
		deprecatedJsonCol: JSONValue | null
		deprecatedUuidCol: string | null
	}
	hasOne: {
	}
	hasMany: {
	}
	hasManyBy: {
	}
}

export type ContemberClientEntities = {
	Foo: Foo
}

export type ContemberClientSchema = {
	entities: ContemberClientEntities
}
"
`;

exports[`generate entities generate for has enum 1`] = `
"import type { FooEnumCol } from './enums'
import type { FooDeprecatedEnumCol } from './enums'
import type { DeprecatedFooEnumCol } from './enums'
import type { DeprecatedFooDeprecatedEnumCol } from './enums'

export type JSONPrimitive = string | number | boolean | null
export type JSONValue = JSONPrimitive | JSONObject | JSONArray
export type JSONObject = { readonly [K in string]?: JSONValue }
export type JSONArray = readonly JSONValue[]

export type Foo <OverRelation extends string | never = never> = {
	name: 'Foo'
	unique:
		| Omit<{ id: string}, OverRelation>
	columns: {
		id: string
		enumCol: FooEnumCol | null
	}
	hasOne: {
	}
	hasMany: {
	}
	hasManyBy: {
	}
}

export type ContemberClientEntities = {
	Foo: Foo
}

export type ContemberClientSchema = {
	entities: ContemberClientEntities
}
"
`;

exports[`generate entities generate for has enum deprecated 1`] = `
"import type { FooEnumCol } from './enums'
import type { FooDeprecatedEnumCol } from './enums'
import type { DeprecatedFooEnumCol } from './enums'
import type { DeprecatedFooDeprecatedEnumCol } from './enums'

export type JSONPrimitive = string | number | boolean | null
export type JSONValue = JSONPrimitive | JSONObject | JSONArray
export type JSONObject = { readonly [K in string]?: JSONValue }
export type JSONArray = readonly JSONValue[]

export type Foo <OverRelation extends string | never = never> = {
	name: 'Foo'
	unique:
		| Omit<{ id: string}, OverRelation>
	columns: {
		id: string
		enumCol: FooEnumCol | null
		deprecatedEnumCol: FooDeprecatedEnumCol | null
	}
	hasOne: {
	}
	hasMany: {
	}
	hasManyBy: {
	}
}
export type DeprecatedFoo <OverRelation extends string | never = never> = {
	name: 'DeprecatedFoo'
	unique:
		| Omit<{ id: string}, OverRelation>
	columns: {
		id: string
		enumCol: DeprecatedFooEnumCol | null
		deprecatedEnumCol: DeprecatedFooDeprecatedEnumCol | null
	}
	hasOne: {
	}
	hasMany: {
	}
	hasManyBy: {
	}
}

export type ContemberClientEntities = {
	Foo: Foo
	DeprecatedFoo: DeprecatedFoo
}

export type ContemberClientSchema = {
	entities: ContemberClientEntities
}
"
`;

exports[`generate entities generate one has one 1`] = `
"
export type JSONPrimitive = string | number | boolean | null
export type JSONValue = JSONPrimitive | JSONObject | JSONArray
export type JSONObject = { readonly [K in string]?: JSONValue }
export type JSONArray = readonly JSONValue[]

export type Foo <OverRelation extends string | never = never> = {
	name: 'Foo'
	unique:
		| Omit<{ id: string}, OverRelation>
		| Omit<{ oneHasOneInverseRel: Bar['unique']}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
		oneHasOneInverseRel: Bar
	}
	hasMany: {
	}
	hasManyBy: {
	}
}
export type Bar <OverRelation extends string | never = never> = {
	name: 'Bar'
	unique:
		| Omit<{ id: string}, OverRelation>
		| Omit<{ oneHasOneOwningRel: Foo['unique']}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
		oneHasOneOwningRel: Foo
	}
	hasMany: {
	}
	hasManyBy: {
	}
}

export type ContemberClientEntities = {
	Foo: Foo
	Bar: Bar
}

export type ContemberClientSchema = {
	entities: ContemberClientEntities
}
"
`;

exports[`generate entities generate one has one deprecated 1`] = `
"
export type JSONPrimitive = string | number | boolean | null
export type JSONValue = JSONPrimitive | JSONObject | JSONArray
export type JSONObject = { readonly [K in string]?: JSONValue }
export type JSONArray = readonly JSONValue[]

export type Foo <OverRelation extends string | never = never> = {
	name: 'Foo'
	unique:
		| Omit<{ id: string}, OverRelation>
		| Omit<{ oneHasOneInverseRel: Bar['unique']}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
		oneHasOneInverseRel: Bar
	}
	hasMany: {
	}
	hasManyBy: {
	}
}
export type Bar <OverRelation extends string | never = never> = {
	name: 'Bar'
	unique:
		| Omit<{ id: string}, OverRelation>
		| Omit<{ oneHasOneOwningRel: Foo['unique']}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
		oneHasOneOwningRel: Foo
	}
	hasMany: {
	}
	hasManyBy: {
	}
}
export type DeprecatedFoo <OverRelation extends string | never = never> = {
	name: 'DeprecatedFoo'
	unique:
		| Omit<{ id: string}, OverRelation>
		| Omit<{ deprecatedOneHasOneInverseRel: DeprecatedBar['unique']}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
		deprecatedOneHasOneInverseRel: DeprecatedBar
	}
	hasMany: {
	}
	hasManyBy: {
	}
}
export type DeprecatedBar <OverRelation extends string | never = never> = {
	name: 'DeprecatedBar'
	unique:
		| Omit<{ id: string}, OverRelation>
		| Omit<{ deprecatedOneHasOneOwningRel: DeprecatedFoo['unique']}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
		deprecatedOneHasOneOwningRel: DeprecatedFoo
	}
	hasMany: {
	}
	hasManyBy: {
	}
}

export type ContemberClientEntities = {
	Foo: Foo
	Bar: Bar
	DeprecatedFoo: DeprecatedFoo
	DeprecatedBar: DeprecatedBar
}

export type ContemberClientSchema = {
	entities: ContemberClientEntities
}
"
`;

exports[`generate entities generate one has many 1`] = `
"
export type JSONPrimitive = string | number | boolean | null
export type JSONValue = JSONPrimitive | JSONObject | JSONArray
export type JSONObject = { readonly [K in string]?: JSONValue }
export type JSONArray = readonly JSONValue[]

export type Foo <OverRelation extends string | never = never> = {
	name: 'Foo'
	unique:
		| Omit<{ id: string}, OverRelation>
		| Omit<{ oneHasManyRel: Bar['unique']}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
	}
	hasMany: {
		oneHasManyRel: Bar<'manyHasOneRel'>
	}
	hasManyBy: {
	}
}
export type Bar <OverRelation extends string | never = never> = {
	name: 'Bar'
	unique:
		| Omit<{ id: string}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
		manyHasOneRel: Foo
	}
	hasMany: {
	}
	hasManyBy: {
	}
}

export type ContemberClientEntities = {
	Foo: Foo
	Bar: Bar
}

export type ContemberClientSchema = {
	entities: ContemberClientEntities
}
"
`;

exports[`generate entities generate one has many deprecated 1`] = `
"
export type JSONPrimitive = string | number | boolean | null
export type JSONValue = JSONPrimitive | JSONObject | JSONArray
export type JSONObject = { readonly [K in string]?: JSONValue }
export type JSONArray = readonly JSONValue[]

export type Foo <OverRelation extends string | never = never> = {
	name: 'Foo'
	unique:
		| Omit<{ id: string}, OverRelation>
		| Omit<{ oneHasManyRel: Bar['unique']}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
	}
	hasMany: {
		oneHasManyRel: Bar<'manyHasOneRel'>
	}
	hasManyBy: {
	}
}
export type Bar <OverRelation extends string | never = never> = {
	name: 'Bar'
	unique:
		| Omit<{ id: string}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
		manyHasOneRel: Foo
	}
	hasMany: {
	}
	hasManyBy: {
	}
}
export type DeprecatedFoo <OverRelation extends string | never = never> = {
	name: 'DeprecatedFoo'
	unique:
		| Omit<{ id: string}, OverRelation>
		| Omit<{ deprecatedOneHasManyRel: DeprecatedBar['unique']}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
	}
	hasMany: {
		deprecatedOneHasManyRel: DeprecatedBar<'deprecatedManyHasOneRel'>
	}
	hasManyBy: {
	}
}
export type DeprecatedBar <OverRelation extends string | never = never> = {
	name: 'DeprecatedBar'
	unique:
		| Omit<{ id: string}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
		deprecatedManyHasOneRel: DeprecatedFoo
	}
	hasMany: {
	}
	hasManyBy: {
	}
}

export type ContemberClientEntities = {
	Foo: Foo
	Bar: Bar
	DeprecatedFoo: DeprecatedFoo
	DeprecatedBar: DeprecatedBar
}

export type ContemberClientSchema = {
	entities: ContemberClientEntities
}
"
`;

exports[`generate entities generate many has many 1`] = `
"
export type JSONPrimitive = string | number | boolean | null
export type JSONValue = JSONPrimitive | JSONObject | JSONArray
export type JSONObject = { readonly [K in string]?: JSONValue }
export type JSONArray = readonly JSONValue[]

export type Foo <OverRelation extends string | never = never> = {
	name: 'Foo'
	unique:
		| Omit<{ id: string}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
	}
	hasMany: {
		manyHasManyRel: Bar
	}
	hasManyBy: {
	}
}
export type Bar <OverRelation extends string | never = never> = {
	name: 'Bar'
	unique:
		| Omit<{ id: string}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
	}
	hasMany: {
		manyHasManyInverseRel: Foo
	}
	hasManyBy: {
	}
}

export type ContemberClientEntities = {
	Foo: Foo
	Bar: Bar
}

export type ContemberClientSchema = {
	entities: ContemberClientEntities
}
"
`;

exports[`generate entities generate many has many deprecated 1`] = `
"
export type JSONPrimitive = string | number | boolean | null
export type JSONValue = JSONPrimitive | JSONObject | JSONArray
export type JSONObject = { readonly [K in string]?: JSONValue }
export type JSONArray = readonly JSONValue[]

export type Foo <OverRelation extends string | never = never> = {
	name: 'Foo'
	unique:
		| Omit<{ id: string}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
	}
	hasMany: {
		manyHasManyRel: Bar
	}
	hasManyBy: {
	}
}
export type Bar <OverRelation extends string | never = never> = {
	name: 'Bar'
	unique:
		| Omit<{ id: string}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
	}
	hasMany: {
		manyHasManyInverseRel: Foo
	}
	hasManyBy: {
	}
}
export type DeprecatedFoo <OverRelation extends string | never = never> = {
	name: 'DeprecatedFoo'
	unique:
		| Omit<{ id: string}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
	}
	hasMany: {
		deprecatedManyHasManyRel: DeprecatedBar
	}
	hasManyBy: {
	}
}
export type DeprecatedBar <OverRelation extends string | never = never> = {
	name: 'DeprecatedBar'
	unique:
		| Omit<{ id: string}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
	}
	hasMany: {
		deeprecatedManyHasManyInverseRel: DeprecatedFoo
	}
	hasManyBy: {
	}
}

export type ContemberClientEntities = {
	Foo: Foo
	Bar: Bar
	DeprecatedFoo: DeprecatedFoo
	DeprecatedBar: DeprecatedBar
}

export type ContemberClientSchema = {
	entities: ContemberClientEntities
}
"
`;

exports[`generate entities generate reduced has by 1`] = `
"
export type JSONPrimitive = string | number | boolean | null
export type JSONValue = JSONPrimitive | JSONObject | JSONArray
export type JSONObject = { readonly [K in string]?: JSONValue }
export type JSONArray = readonly JSONValue[]

export type Foo <OverRelation extends string | never = never> = {
	name: 'Foo'
	unique:
		| Omit<{ id: string}, OverRelation>
		| Omit<{ locales: FooLocale['unique']}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
	}
	hasMany: {
		locales: FooLocale<'foo'>
	}
	hasManyBy: {
		localesByLocale: { entity: FooLocale; by: {locale: string}  }
	}
}
export type FooLocale <OverRelation extends string | never = never> = {
	name: 'FooLocale'
	unique:
		| Omit<{ id: string}, OverRelation>
		| Omit<{ locale: string; foo: Foo['unique']}, OverRelation>
	columns: {
		id: string
		locale: string
	}
	hasOne: {
		foo: Foo
	}
	hasMany: {
	}
	hasManyBy: {
	}
}

export type ContemberClientEntities = {
	Foo: Foo
	FooLocale: FooLocale
}

export type ContemberClientSchema = {
	entities: ContemberClientEntities
}
"
`;

exports[`generate entities generate reduced has by deprecated 1`] = `
"
export type JSONPrimitive = string | number | boolean | null
export type JSONValue = JSONPrimitive | JSONObject | JSONArray
export type JSONObject = { readonly [K in string]?: JSONValue }
export type JSONArray = readonly JSONValue[]

export type Foo <OverRelation extends string | never = never> = {
	name: 'Foo'
	unique:
		| Omit<{ id: string}, OverRelation>
		| Omit<{ locales: FooLocale['unique']}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
	}
	hasMany: {
		locales: FooLocale<'foo'>
	}
	hasManyBy: {
		localesByLocale: { entity: FooLocale; by: {locale: string}  }
	}
}
export type FooLocale <OverRelation extends string | never = never> = {
	name: 'FooLocale'
	unique:
		| Omit<{ id: string}, OverRelation>
		| Omit<{ locale: string; foo: Foo['unique']}, OverRelation>
	columns: {
		id: string
		locale: string
	}
	hasOne: {
		foo: Foo
	}
	hasMany: {
	}
	hasManyBy: {
	}
}
export type DeprecatedFoo <OverRelation extends string | never = never> = {
	name: 'DeprecatedFoo'
	unique:
		| Omit<{ id: string}, OverRelation>
		| Omit<{ deprecatedLocales: DeprecatedFooLocale['unique']}, OverRelation>
	columns: {
		id: string
	}
	hasOne: {
	}
	hasMany: {
		deprecatedLocales: DeprecatedFooLocale<'deprecatedFoo'>
	}
	hasManyBy: {
	}
}
export type DeprecatedFooLocale <OverRelation extends string | never = never> = {
	name: 'DeprecatedFooLocale'
	unique:
		| Omit<{ id: string}, OverRelation>
		| Omit<{ deprecatedLocale: string; deprecatedFoo: DeprecatedFoo['unique']}, OverRelation>
	columns: {
		id: string
		deprecatedLocale: string
	}
	hasOne: {
		deprecatedFoo: DeprecatedFoo
	}
	hasMany: {
	}
	hasManyBy: {
	}
}

export type ContemberClientEntities = {
	Foo: Foo
	FooLocale: FooLocale
	DeprecatedFoo: DeprecatedFoo
	DeprecatedFooLocale: DeprecatedFooLocale
}

export type ContemberClientSchema = {
	entities: ContemberClientEntities
}
"
`;
